### PR TITLE
Allow prange and parallel sections without `nogil`

### PR DESCRIPTION
**Pull request by da-woods**
This is mostly targeted at freethreading builds.

It also works on GIL-builds and specifically optimises the following pattern to avoid costly GIL juggling:
```cython
for i in prange(...):
  with gil:
    ...
```
Adds a warning when the GIL is not released in a build that is not marked as freethreading-compatible.

### CODE REVIEW

File: Cython/Compiler/Nodes.py [hunk 1]
```diff
@@ -9474,6 +9474,19 @@ def analyse_declarations(self, env):
...
+            # explicitly release or acquire the GIL.
+            error(self.pos, "Contradictory settings for 'with_gil' and 'nogil'")
+        if self.with_gil:
+            # For now warn against this pretty heavily. I don't want to deal with too
+            # many bug reports at this stage.
+            warning(
+                self.pos,
+                "'with_gil' is experimental. Cython does almost nothing to ensure "
+                "thread safe use of Python variables (including reference counting) "
+                "and so it is completely your responsibility. Use with caution!", 999)
```
Comment (da-woods) [inline Cython/Compiler/Nodes.py:9489]: I don't intend to keep this warning forever. But I think we need more control of PyObject variables in parallel blocks before we can really start recommending it.

File: Cython/Compiler/Nodes.py [hunk 2]
```diff
@@ -10472,8 +10508,39 @@ def generate_loop(self, code, fmt_dict):
                 code.putln("#if 0")
             else:
                 code.putln("#ifdef _OPENMP")
+            with_gil_deadlock_avoidance_point = code.insertion_point()
             code.put("#pragma omp for")
 
+        if self.with_gil:
+            # Any firstprivate creates a barrier at least on GCC (and thus
+            # a deadlock if we're using the GIL). And there's also an implicit
```
Comment (lysnikolaou) [inline Cython/Compiler/Nodes.py:10516]: Have you examined whether this would also be relevant if someone's using critical sections inside an openmp loop? If so, should we document that?

File: docs/src/userguide/parallelism.rst [hunk 3]
```diff
@@ -12,15 +12,26 @@ Using Parallelism
...
 
 .. NOTE:: Functionality in this module may only be used from the main thread
           or parallel regions due to OpenMP restrictions.
 
 
-.. function:: prange([start,] stop[, step][, nogil=False][, use_threads_if=CONDITION][, schedule=None[, chunksize=None]][, num_threads=None])
+Historically, this kind of parallelism could only be used with the
+:term:`GIL<Global Interpreter Lock or GIL>` released (see :ref:`Releasing the GIL <nogil>`).
+However, there is now some experimental support for running with the GIL in
+freethreading builds. You must either use these functions in a no-gil block,
```
Comment (lysnikolaou) [inline docs/src/userguide/parallelism.rst:24]: I guess that this wording will be confusing for most people that are not intimately familiar with how the free-threaded build works internally. Maybe we could reword it to something else like "with the thread state saved"? Although that seems suboptimal as well.

Ditto for other places.

**Review by lysnikolaou — COMMENTED** [review]
Left a couple of questions. Maybe also the name of the paremeter should be changed from `with_gil` to something more hinting at the free-threaded build?

File: Cython/Compiler/Nodes.py (hunk 2 above)
Comment (da-woods) [reply Cython/Compiler/Nodes.py:10516]: I'm assuming you mean `with cython.critical_section`? (I ask because OpenMP has a "critical_section" feature that we use internally to guard certain small blocks... I think we use that safely though).

I don't think that's `cython.critical_section` should be an issue here. The barrier is just that it waits for all threads to be ready to go before the absolutely first iteration.  So it's just trying not to hold the GIL when that happens. There shouldn't be an issue with anyone using critical sections or other kinds of locks within the loop itself.

File: docs/src/userguide/parallelism.rst (hunk 3 above)
Comment (da-woods) [reply docs/src/userguide/parallelism.rst:24]: Yes good point.

I think we're stuck with the naming for the existing `with gil:`/`with nogil` features. But I can add an explanation if the free-threading docs about what that means the the context of a freethreading build.

But since this is a new feature I can probably pick a different name for the parameter. And I'll try to rephrase the docs to be a better clearer.

**Comment by da-woods** [comment]
I've renamed `with_gil` in this context to `with_python` and tried to clarify the docs here.

**Comment by lysnikolaou** [comment]
> I've renamed `with_gil` in this context to `with_python` and tried to clarify the docs here.

That sounds better. What do you think about something like `with_thread_state_saved`?

**Comment by da-woods** [comment]
> > I've renamed `with_gil` in this context to `with_python` and tried to clarify the docs here.
> 
> That sounds better. What do you think about something like `with_thread_state_saved`?

Probably too technical. I assume most users would rather not know about thread state at all.  (Although I know the existing `with gil`/`with nogil` are similarly technical).

I prefer `with_python` because it really just describes what you can do.

File: Cython/Compiler/Nodes.py [hunk 4]
```diff
@@ -9911,22 +9928,35 @@ def end_parallel_block(self, code):
...
+                end_code.putln("#ifdef _OPENMP")
+                end_code.putln("{")
+                end_code.putln("int had_gil = PyGILState_Check();")
+                end_code.putln("PyThreadState *_save;")
+                end_code.putln("if (had_gil) Py_UNBLOCK_THREADS")
+                # Ensure the GIL is unblocked, and wait for all threads before proceeding.
+                end_code.putln("#pragma omp barrier")
+                end_code.putln("if (had_gil) Py_BLOCK_THREADS")
+                end_code.putln("}")
+                end_code.putln("#endif /* _OPENMP */")
```
Comment (da-woods) [inline Cython/Compiler/Nodes.py:9959]: This wants to use https://github.com/cython/cython/pull/6847 eventually

**Comment by scoder** [comment]
So, I looked over this and read through the documentation snippets, but I still feel unsure about the goal and effects of this change. I gathered that the new `with_python=True` option assures a valid thread state for the OpenMP threads but does not itself control the GIL state, is that correct? Why can't all OpenMP threads have Python thread states right away, without a special option?

Could you maybe extend the PR description with some kind of problem description and motivation?

**Comment by da-woods** [comment]
It's pretty close to writing

```
for i in prange(N, nogil=True):
    with gil:
        ... # code
```

It's _slightly_ more efficient (just by not using the `PyGILState` API as much) but not as much as I hoped when I first started looking at it - the big limitation is that Cython does have to release and re-acquire the GIL each time you go round the loop to avoid deadlock (or somehow eliminate `last_private` from the loop, which is is what requires the barrier).

> Why can't all OpenMP threads have Python thread states right away, without a special option?

Mainly because it's a bit of a pitfall with the non-freethreading interpreter. It means that people that accidentally omit `nogil=True` will not get any parallelism even if they aren't using the GIL. So at least for me it seems better to make it explicit

**Comment by da-woods** [comment]
I'm possibly not explaining very well

```
for i in prange(N, with_python=True):
    ... # code
```

is transformed to something pretty close to

```
for i in prange(N, nogil=True):
    with gil:
        ... # code

```

(but a little faster)

> > people that accidentally omit `nogil=True` will not get any parallelism even if they aren't using the GIL
> 
> But isn't that expected? I think the documentation states pretty clearly that parallelism requires releasing the GIL. Wasn't it always the case that you get sequential "threads" if you run a `prange` loop without releasing the GIL?

Right now you get a compile-time error if you try to write a `prange` loop without releasing the GIL.

-----------------------------------------------------------------------------------------

But yes - in principle we could just try to speed up `prange(N, nogil=True): with gil:` instead and just document that as the idiomatic approach. Given that it doesn't seem possible to safely go round the loop without releasing the GIL the benefit of this PR is fairly minor anyway

**Comment by scoder** [comment]
> Right now you get a compile-time error if you try to write a `prange` loop without releasing the GIL.

Ah, so, we basically push people into releasing the GIL even if they directly need it afterwards, is that it? Ok, that use case doesn't seem hugely important to me – if you want parallelism, you probably want it for parallel code, not sequential code. But I can see that it can be useful to do some kind of Python level initialisation inside of the thread first. We also have parallel sections for stuff like that, but well, it's still a use case. And I think the obvious idiom for it is `for in prange: with gil:`. If there's a way to optimise such code, why not. I don't see why we should bother users with yet another option.

Do you have an example of real world code where this is used?

**Comment by scoder** [comment]
BTW, we don't have any benchmarks for our parallel code, neither micro nor macro. Since you're working on the freethreading and parallel code, maybe you could come up with at least some microbenchmarks for this kind of mixture between Python and nogil code? I'm not sure if they'll run well in CI (could be that we have a single CPU core there), but we'll see. At least sequential locking and critical sections might still be candidates for measuring their overhead.

**Comment by da-woods** [comment]
>  Ah, so, we basically push people into releasing the GIL even if they directly need it afterwards, is that it? Ok, that use case doesn't seem hugely important to me – if you want parallelism, you probably want it for parallel code, not sequential code

The use case is that in freethreaded Python this code is parallel.

------------------------------------------

This is mainly targeted at people using freethreaded Python who want to use `prange` loops without manually messing around with the GIL. (On non-freethreaded Python the intention is that it works, but clearly the GIL means it won't work well).

The main limitation is that it isn't currently much better than writing `for i in prange(N, nogil=True): with gil: ...`, so it's mainly a better expression of intent than better C code.

I subsequently had a thought about how to avoid needing to release the GIL on each loop (at least in some cases) so maybe I should look at that first - it's easier to justify something new if it genuinely does lead to better  performance

**Comment by da-woods** [comment]
So I think I've been able to drop the "release the GIL on every go round", which should be an improvement.

A quick timing test:

```
# cython: freethreading_compatible=True

# distutils: extra_compile_args=-fopenmp
# distutils: extra_link_args=-fopenmp

from cython.parallel import prange

def f(i):
    return i

def test_prange_with_python(int N):
    cdef int i = 0

    out = [None]*N

    for i in prange(N, with_python=True):
        out[i] = f(i)
    
def test_old_way(int N):
    cdef int i = 0

    out = [None]*N

    for i in prange(N, nogil=True):
        with gil:
            out[i] = f(i)

def do_timing(N):
    from datetime import datetime
    t0 = datetime.now()
    test_old_way(N)
    t1 = datetime.now()
    test_prange_with_python(N)
    t2 = datetime.now()

    print(f"old way:", t1-t0)
    print(f"new way:", t2-t1)
```

On freethreading Python 3.13 `N=1000000` the `with_python` version is consistently about 10% faster (i.e. noticeable but not huge):

```
old way: 0:00:00.249138
new way: 0:00:00.221372
```

On GIL Python 3.13 `N=1000000` the `with_python` version is dramatically faster 

```
old way: 0:00:28.616594
new way: 0:00:00.073102
```

(i.e. when the GIL is contended then constantly swapping it is absolutely awful).

-------------------------------------------------------

So the upshot is, it's a mild improvement for the freethreading builds (for which it's intended) and a dramatic improvement for the GIL build (where it isn't intended)

**Comment by scoder** [comment]
> freethreading_compatible=True

Isn't this a good indicator for whether we should warn about the GIL being released or not when we enter a prange loop? Even with your speed improvement, it still matters in non-FT builds, right?

**Comment by da-woods** [comment]
> Isn't this a good indicator for whether we should warn about the GIL being released or not when we enter a prange loop?

I think what you're suggesting is:
* Remove the explicit `with_gil` directive,
* If the user omits `nogil=True` then don't error - let them use the GIL.
* Warn if they don't release the GIL and don't mark it as `freethreading_compatible`

?

If so, that seems reasonable to me.

> Even with your speed improvement, it still matters in non-FT builds, right?

Yes - on non-FT builds it's running sequentially, just not fighting over the GIL quite so often.

**Comment by scoder** [comment]
> * Remove the explicit `with_gil` directive,
> * If the user omits `nogil=True` then don't error - let them use the GIL.
> * Warn if they don't release the GIL and don't mark it as `freethreading_compatible`

Yes, that's what I was thinking.

File: tests/errors/e_cython_parallel.pyx [hunk 5]
```diff
@@ -171,41 +175,54 @@ def bar():
...
-170:49: use_threads_if may not be a Python object as we don't have the GIL
+7:8: cython.parallel.parallel is not a module
+8:0: No such directive: cython.parallel.something
+10:7: cython.parallel.parallel is not a module
+11:0: No such directive: cython.parallel.something
+17:6: prange() can only be used as part of a for loop
+17:6: prange without releasing the GIL will only work well on freethreaded Python
+22:19: Invalid schedule argument to prange: invalid_schedule
+25:29: Parallel section without releasing the GIL will only work well on freethreaded Python
+31:8: target may not be a Python object as we don't have the GIL
```
Comment (da-woods) [inline tests/errors/e_cython_parallel.pyx:187]: Arguable some of these restrictions could be relaxed (or the error message updated). But that's probably best done in a follow-up at some point

File: Cython/Compiler/Nodes.py [hunk 6]
```diff
@@ -10726,6 +10750,16 @@ def generate_loop(self, code, fmt_dict):
         code.end_block()  # end guard around loop body
         code.end_block()  # end for loop block
 
+        if self.acquire_gil:
+            code.putln("#ifdef _OPENMP")
+            code.putln("if (!__pyx_parallel_loop_threadstate) {")
+            code.putln("__pyx_parallel_loop_threadstate = PyEval_SaveThread();")
```
Comment (scoder) [inline Cython/Compiler/Nodes.py]: -> `Naming.py` ?

File: docs/src/userguide/freethreading.rst [hunk 7]
```diff
@@ -51,6 +51,20 @@ extension modules claim to support it then you can either:
 
 These options are mainly useful for testing.
 
+Understanding the "GIL" in free-threaded Python
+===============================================
+
+Although free-threading builds don't have a Global Interpreter Lock, they do still
+need to have a suitable Python "thread-state" when running code that touches Python
+objects.  When the Cython documentation refers to "holding the GIL" you should
+interpret it "being in a state where you can run Python code", rather than holding
```
Comment (scoder) [inline docs/src/userguide/freethreading.rst]:
```suggestion
interpret it as "being in a state where you can run Python code", rather than holding
```

File: docs/src/userguide/freethreading.rst [hunk 8]
```diff
@@ -51,6 +51,20 @@ extension modules claim to support it then you can either:
...
+an exclusive lock.
+
+In Cython that thread-state is still controlled by the existing ``with gil:`` and
+``with nogil:`` context managers.  There may still be a value in placing code that doesn't need
+to use Python objects in a ``with nogil`` so that it works well in the regular
```
Comment (scoder) [inline docs/src/userguide/freethreading.rst]:
```suggestion
to use Python objects in a ``with nogil`` block so that it works well in the regular
```

File: docs/src/userguide/parallelism.rst [hunk 9]
```diff
@@ -12,14 +12,26 @@ Using Parallelism
...
-It currently supports OpenMP, but later on more backends might be supported.
+module.  It currently supports OpenMP, but later on more backends might be supported.
 
 .. NOTE:: Functionality in this module may only be used from the main thread
           or parallel regions due to OpenMP restrictions.
 
 
+Historically, this kind of parallelism could only be used with the
+:term:`GIL<Global Interpreter Lock or GIL>` released (see :ref:`Releasing the GIL <nogil>`).
+However, from Cython 3.2 there is now some experimental support for running these parallel blocks with
```
Comment (scoder) [inline docs/src/userguide/parallelism.rst]:
```suggestion
However, from Cython 3.3 there is now some experimental support for running these parallel blocks with
```

**Review by scoder — COMMENTED** [review]
The PR title and description (i.e. the future commit comment) now needs adapting, but I'd then be happy to see this merged.
